### PR TITLE
python3Packages.numpy: mass-rebuild clean up

### PR DESCRIPTION
**Pull request by vcunat**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/numpy/default.nix [hunk 1]
```diff
@@ -56,16 +56,14 @@ in buildPythonPackage (rec {
     # Patching of numpy.distutils is needed to prevent it from undoing the
     # patch to distutils.
     ./numpy-distutils-C++.patch
-  ]
-  ++ lib.optionals stdenv.cc.isClang [
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:59]: This may work but it is probably wrong. We definitely must do a cleanup PR after this.

Comment (vcunat) [reply pkgs/development/python-modules/numpy/default.nix:59]: I don't understand you.  I thought this PR is the cleanup you wanted.

Comment (reckenrode) [reply pkgs/development/python-modules/numpy/default.nix:59]: If we’re removing the conditionals, shouldn’t the x86_64-darwin one also be removed?

Comment (vcunat) [reply pkgs/development/python-modules/numpy/default.nix:59]: I considered that (as seen in the commit message), but I didn't expect we want to drop the test on other platforms.  And AFAIK that patch wasn't upstreamed, so the likelihood seems lower for the usual clash on update where reviewers don't drop a platform-specific patch.

Comment (reckenrode) [reply pkgs/development/python-modules/numpy/default.nix:59]: The patch uses `pytest.mark.skipif` conditional on platform, but if it’s also conditional in the derivation, that’s fine. That was just raised as a concern in the review of the other PR.

Comment (vcunat) [reply pkgs/development/python-modules/numpy/default.nix:59]: Oh, I missed that.  So applied unconditionally now.

Comment (SuperSandro2000) [reply pkgs/development/python-modules/numpy/default.nix]: I think I mistook this PR for the original and not the cleanup, upsi, sorry

File: pkgs/development/python-modules/numpy/default.nix [hunk 2]
```diff
@@ -91,6 +89,9 @@ in buildPythonPackage (rec {
   nativeBuildInputs = [ gfortran cython ];
   buildInputs = [ blas lapack ];
 
+  # Causes `error: argument unused during compilation: '-fno-strict-overflow'` due to `-Werror`.
+  hardeningDisable = lib.optionals stdenv.cc.isClang [ "strictoverflow" ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:93]: We can't set the warn-error flag from here https://github.com/numpy/numpy/blob/0aaa5d397f86d5b4aef9c83053f65296a790e501/numpy/distutils/command/build.py#L17 somehow?

File: pkgs/development/python-modules/numpy/default.nix [hunk 3]
```diff
@@ -51,33 +51,31 @@ in buildPythonPackage (rec {
...
       hash = "sha256-cxBoimX5a9wC2qUIGAo5o/M2E9+eV63bV2/wLmfDYKg=";
     })
-  ]
-  ++ lib.optionals (stdenv.isDarwin && stdenv.isx86_64) [
+
     # Disable `numpy/core/tests/test_umath.py::TestComplexFunctions::test_loss_of_precision[complex256]`
     # on x86_64-darwin because it fails under Rosetta 2 due to issues with trig functions and
     # 80-bit long double complex numbers.
     ./disable-failing-long-double-test-Rosetta-2.patch
+  ]
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:73]: that looks good to me

File: pkgs/development/python-modules/numpy/default.nix (hunk 2 above)
Comment (vcunat) [reply pkgs/development/python-modules/numpy/default.nix:93]: That's more of a discussion for the original PR.  Here I just move the line to simplify the expression.

I would expect that some `-Wno-error` approach would also work.

Comment (reckenrode) [reply pkgs/development/python-modules/numpy/default.nix:93]: Doesn’t it make more sense to stop passing the unused argument than to hack at the build scripts to suppress the unused argument warning?

See #243595 regarding `-fno-strict-overflow` and Clang.

Comment (SuperSandro2000) [reply pkgs/development/python-modules/numpy/default.nix:93]: Ah, sorry.

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**